crypto/thread/api.c

```diff
@@ -0,0 +1,265 @@
+    tdata->enabled = 1;
+    return 1;
+}
+
+int CRYPTO_THREAD_disable(OPENSSL_CTX *ctx)
+{
+    OPENSSL_CTX_THREADS tdata;
+    CRYPTO_WORKER worker;
+    struct list *iter, *tmp;
+
+    tdata = OPENSSL_CTX_THREADS(ctx);
+    if (tdata == NULL)
+        return 0;
+
+    CRYPTO_MUTEX_lock(tdata->lock);
+    tdata->threads.cap = 0;
+    tdata->enabled = 0;
+
+    list_for_each_safe(iter, tmp, &tdata->workers.available) {
+        worker = container_of(iter, struct crypto_worker_st, list);
+        CRYPTO_THREAD_native_terminate(worker->handle);
+        CRYPTO_THREAD_native_clean(worker->handle);
+        list_del(iter);
+        OPENSSL_free(worker);
+    }
+    list_for_each_safe(iter, tmp, &tdata->workers.busy) {
+        worker = container_of(iter, struct crypto_worker_st, list);
+        CRYPTO_THREAD_native_terminate(worker->handle);
+        CRYPTO_THREAD_native_clean(worker->handle);
+        list_del(iter);
+        OPENSSL_free(worker);
+    }
+    list_for_each_safe(iter, tmp, &tdata->threads.active) {
+        worker = container_of(iter, struct crypto_worker_st, list);
+        CRYPTO_THREAD_native_terminate(worker->handle);
+        CRYPTO_THREAD_native_clean(worker->handle);
+        list_del(iter);
+        OPENSSL_free(worker);
+    }
+    CRYPTO_MUTEX_unlock(tdata->lock);
+    CRYPTO_THREAD_clean(ctx, NULL);
+    return 1;
+}
+
+int CRYPTO_THREAD_cap(OPENSSL_CTX *ctx, int max_threads)
+{
+    OPENSSL_CTX_THREADS tdata;
+
+    tdata = OPENSSL_CTX_THREADS(ctx);
+    if (tdata == NULL)
+        return 0;
+
+    if (tdata->enabled == 0)
+        return 0;
+
+    tdata->threads.cap = max_threads;
+    return 1;
+}
+
+size_t CRYPTO_THREAD_get_available_threads(OPENSSL_CTX *ctx)
+{
+    OPENSSL_CTX_THREADS t;
+
+    t = OPENSSL_CTX_THREADS(ctx);
+    if (t == NULL)
+        return 0;
+
+    if (t->enabled == 0)
+        return 0;
+
+    if (t->threads.cap < 0)
+        return -1;
+
+    return t->threads.cap - list_size(&t->threads.active) +
+        list_size(&t->workers.available);
+}
+
+static int CRYPTO_THREAD_spawn_worker_task(OPENSSL_CTX *ctx,
+                                           CRYPTO_WORKER_CALLBACK cb,
+                                           void *vtask)
+{
+    int available_threads;
+    CRYPTO_WORKER worker;
+    OPENSSL_CTX_THREADS tdata;
+
+    tdata = OPENSSL_CTX_THREADS(ctx);
+    if (tdata == NULL)
+        return 0;
+
+    if (tdata->enabled == 0)
+        return 0;
+
+    worker = CRYPTO_WORKER_new(ctx, cb, vtask);
+    if (worker == NULL)
+        goto fail;
+
+    if (vtask != NULL)
+        list_add_tail(&worker->list, &tdata->threads.active);
+
+    available_threads = CRYPTO_THREAD_get_available_threads(ctx);
+    worker->handle = CRYPTO_THREAD_native_start(worker_main, (void*)worker, 0);
+    if (worker->handle == NULL) {
+        list_del(&worker->list);
+        goto fail;
+    }
+
+    while (vtask == NULL &&
+           CRYPTO_THREAD_get_available_threads(ctx) == available_threads)
```

Comment:
A worker that stops and does not poll for new task does not influence this in any way, so let's assume that a worker stops and tries to poll for a new task. The `vtask == NULL` means that there are no tasks currently available, so any existing or new thread will be blocked -- from that pt. of view it doesn't impact the caller as a worker thread is made available for use, no matter which one. I suppose it could be safer to replace the equality check with inequality to make sure that the loop terminates in the case that both of the 2 threads write into the available list as it's being iterated over by CRYPTO_THREAD_get_available_threads. Provided we don't drop the busywait that is.

Reply:
Changed.